src/analysis_and_optimization/Optimize.ml

```diff
@@ -106,6 +112,20 @@ let subst_args_stmt args es =
   let m = Map.Poly.of_alist_exn (List.zip_exn args es) in
   subst_stmt m
 
+(**
+ * Count the number of returns that happen in a statement
+ *)
+let rec count_returns (acc : int) Stmt.Fixed.{pattern; _} : int =
+  match pattern with
+  | Return _ -> acc + 1
+  | IfElse (_, true_stmt, false_stmt) ->
+      count_returns acc true_stmt
+      + Option.value_map ~default:0 ~f:(count_returns acc) false_stmt
+  | While (_, body) | For {body; _} -> count_returns acc body
+  | Block lst | SList lst | Profile (_, lst) ->
+      List.fold ~init:acc ~f:count_returns lst
+  | _ -> acc
```

Comment:
I think this function could just be a `Stmt.Fixed.fold`?

Reply:
Sorry I think I'm just unfamiliar, what would that look like?

Reply:
Folding functions will walk over a data structure and call a two argument function on each node. In some languages this is called `reduce` instead of `fold`. The first argument will be the "accumulator", the second will be the node. Since our Stmt actually has two types inside (the expression type and then the statement type itself) you give two functions, one for each.

The result is something like  

```ocaml
let count_returns = 
Stmt.Fixed.fold  (fun acc _ -> acc) 
(fun acc -> function Stmt.Fixed.{pattern=Return _; _} -> acc + 1 | _ -> acc) 
0
```
(I don't have the source in front of me to make sure that compiles, but it should be close)

It basically avoids the need for defining lots of functions that look similar to what you've defined that mostly just walk over all the sub nodes. There are a bunch of usages of it around the optimize code like here:

https://github.com/stan-dev/stanc3/blob/d2358be75753fc13825c5fd45e03c99e67d2011c/src/analysis_and_optimization/Factor_graph.ml#L37-L44